### ISSUE TITLE
[Feature Request] Override calculated version

### ISSUE THREAD

**Issue by jftanner** [semantic-release/semantic-release#3368]
### New feature motivation

This is a very niche need but hopefully one that would be fairly easy to implement. Essentially, I want to be able to bypass the version-selecting behavior of semantic-release and instruct it to publish a specific version instead.

> "Why would you want to do that? Doesn't that defeat the entire point of semantic-release?"

Well yes, yes it does. At least partially. But hear me out...

There's ultimately two separate but related tasks that semantic-release performs for me:
1. It selects a new version for that is "unromantic and unsentimental".
2. Is orchestrates the release of that version through whatever plugins I have configured.

Most of the time, I want it to do both of those things. But not _always_.
If I want the first task, but not the second, I can use `--dry-run`. That's pretty easy.
But what if I want the second task without the first? This request is to let me do that.

> "You still haven't said _why_, though..."

My _specific_ use-case is that I'm updating a shared deployment pipeline used by many teams. It doesn't use semantic-release and I want to change that. The existing logic is simple: it loads the major and minor from `package.json`, checks the git tags for matching patches, and releases the next appropriate patch. To bump the major or minor version, devs must manually update the `package.json`.

I can't force other pipeline users to change their workflow, which means I need to support the old version algorithm *and* commit-based versioning at the same time. But once the version is selected, the rest of the process is the same. So, ideally, I'd like to use semantic-release for that orchestration in either case.

One work-around I considered is creating the tag at build-time using the existing logic (since the code already exists to do that), but then semantic-release will just exit because there are no commits since the last tag.

I'm sure there are other use-cases, but that one is mine.

**Tl;dr:** I want to use the release orchestration from semantic-release, while overriding or bypassing its version selection.

### New feature description

Allow a plugin, API call, or CLI command to override the built-in version selection and publish a specific version.

### New feature implementation

Since this is such a niche issue, the feature shouldn't necessarily be built-in to semantic release or an official plugin. Requiring a custom plugin would be perfectly reasonable.

I can see a few possible implementations:
- ~Update the `analyzeCommits` hook to allow a plugin to return a specific version instead of a type.~
- ~Update the `verifyRelease` hook to allow a plugin to change the selected version.~
- Add an option to the JS API to ignore the result from the `analyzeCommits` step and use a specified version.
- ~Add an option to the JS API to force a (re-)release when the current commit has a tag.~

I would be happy to help contribute the feature if the maintainers are willing to accept it.

**Comment by jftanner** [comment]
## `analyzeCommits` hook:
> Update the `analyzeCommits` hook to allow a plugin to return a specific version instead of a type.

Probably not practical. I don't see a good way to pass an explicit version back from there without adding a bunch of unnecessary complexity and touching bits that shouldn't be touched.

Essentially, I would need to:
1. Update the [output validator](https://github.com/semantic-release/semantic-release/blob/202a2f03424d3e40ee7534203f8489ac8edad058/lib/definitions/plugins.js#L19) to accept a version.
1. Update the [post-processor](https://github.com/semantic-release/semantic-release/blob/202a2f03424d3e40ee7534203f8489ac8edad058/lib/definitions/plugins.js#L24-L30) to return a version if the results includes one.
1. Update [index](https://github.com/semantic-release/semantic-release/blob/202a2f03424d3e40ee7534203f8489ac8edad058/index.js#L174-L187) to identify when it's gotten back a version and change the way it creates `nextRelease`. (Including identify what type the release _should_ be.)

That's a whole lot of core functionality that really shouldn't be modified to support such a niche use-case.

## `verifyRelease` hook:
> Update the `verifyRelease` hook to allow a plugin to change the selected version.

For a moment, I thought this might be possible (if very unwise) without any changes to semantic-release. But it's not, since `context` is (wisely) [protected from mutation](https://github.com/semantic-release/semantic-release/blob/202a2f03424d3e40ee7534203f8489ac8edad058/lib/plugins/normalize.js#L37) by plugins.

Implementing this would require a post-processor on the [verifyRelease definition](https://github.com/semantic-release/semantic-release/blob/202a2f03424d3e40ee7534203f8489ac8edad058/lib/definitions/plugins.js#L32-L36) in order to update the release. But, as before, that's way too big of a change for something like this.

## Force version option:
> Add an option to the JS API to ignore the result from the analyzeCommits step and use a specified version.

**This might actually be viable!**

The one good place to interrupt the usual process is inside [getNextVersion](https://github.com/semantic-release/semantic-release/blob/202a2f03424d3e40ee7534203f8489ac8edad058/lib/get-next-version.js), as it is [called](https://github.com/semantic-release/semantic-release/blob/202a2f03424d3e40ee7534203f8489ac8edad058/index.js#L185), right after `nextRelease` is created, but before it gets a gitTag and name.

That function has access to the full context. So I could fairly easy add a `forceVersion` field to `options` and set it from the JS API or CLI (if added as an option there). It would be a fairly small change and easy to both implement and test.

## Force re-publish option:
> Add an option to the JS API to force a (re-)release when the current commit has a tag.

Without even looking at the code, I can think of a number of ways that could pose a problem. So much would need to change to accomodate it that it's not even worth thinking about.

**Comment by babblebey** [comment]
Just write all these functionality into that single plugin or an array of plugins... Then explicitly add it to your configuration in the `plugin` property to override the defaults.

😉

**Comment by jftanner** [comment]
It's a feature request because it's literally not possible with a plugin. ;)

**Comment by travi** [comment]
sorry, but enabling the specification of a different version is too far outside the goals of the project that i dont think this is a feature we could justify making possible. i understand that you are working within restrictions that make something like this desireable, but semantic-release cannot be the tool that handles that situation for you. the best i can recommend is to implement your own wrapper that either calls semantic-release or doesnt

**Comment by jftanner** [comment]
That's a reasonable answer. I appreciate you taking the time to consider it.
